File: remotion/src/config/conflictRules.ts
import rules from './conflict_rules.json';

export const TITLE_BUFFER_FRAMES = rules.title_buffer_frames as number;
export const STEP_WINDOW_BUFFER = rules.step_window_buffer_frames as number;
export const STEP_BEAT_FALLBACK_DURATION = rules.step_beat_fallback_duration_frames as number;
export const COMPOSITED_LAYOUTS = new Set(rules.composited_layouts as string[]);
export const HOOK_FRAME_LIMIT = rules.hook_frame_limit as number;
export const BROLL_RATE_LIMIT_FRAMES = rules.broll_rate_limit_frames as number;
export const BROLL_STAGGER_FRAMES = rules.broll_stagger_frames as number;

const blocking = rules.blocking_shot_types as Record<string, string[]>;

export const blockingTypes = (momentType: keyof typeof blocking | string): Set<string> =>
  new Set(blocking[momentType] ?? blocking.broll ?? ['TITLE_CARD', 'CUSTOM_VISUAL']);

export const STEP_WINDOW_MOMENT_TYPES = new Set(
  (rules.step_window_moment_types as string[]) ?? ['broll', 'fun', 'role', 'social'],
);

export const BROLL_OVERLAP_DEDUP = Boolean(rules.broll_overlap_dedup);
export const BROLL_MAX_PER_STEP = (rules.broll_max_per_step as number) ?? 1;
export const TOAST_BLOCKS_PIP_BROLL_ONLY = Boolean(rules.toast_blocks_pip_broll_only);

File: src/conflict_rules.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

ENGINE_ROOT = Path(__file__).resolve().parent.parent
RULES_PATH = ENGINE_ROOT / "config" / "conflict_rules.json"

_DEFAULTS = {
    "title_buffer_frames": 12,
    "step_window_buffer_frames": 60,
    "step_beat_fallback_duration_frames": 75,
    "hook_frame_limit": 90,
    "broll_stagger_frames": 10,
    "broll_rate_limit_frames": 240,
    "composited_layouts": ["presenter_on_bg", "presenter_cutout", "immersive_flash"],
    "hook_composited_ignore_blocking": ["TITLE_CARD", "STAT_CALLOUT"],
    "blocking_shot_types": {
        "broll": ["TITLE_CARD", "CUSTOM_VISUAL"],
        "fun": ["TITLE_CARD", "CUSTOM_VISUAL"],
        "role": ["TITLE_CARD", "CUSTOM_VISUAL"],
        "logo": ["TITLE_CARD", "CUSTOM_VISUAL"],
        "social": ["TITLE_CARD", "CUSTOM_VISUAL"],
    },
    "step_window_moment_types": ["broll", "fun", "role", "social"],
    "broll_overlap_dedup": True,
    "broll_max_per_step": 1,
    "toast_blocks_pip_broll_only": True,
}


@lru_cache(maxsize=1)
def load_rules() -> dict:
    if RULES_PATH.exists():
        data = json.loads(RULES_PATH.read_text(encoding="utf-8"))
        merged = dict(_DEFAULTS)
        merged.update(data)
        return merged
    return dict(_DEFAULTS)


def title_buffer_frames() -> int:
    return int(load_rules().get("title_buffer_frames", 12))


def step_window_buffer_frames() -> int:
    return int(load_rules().get("step_window_buffer_frames", 60))


def blocking_types(moment_type: str = "broll") -> set[str]:
    rules = load_rules()
    types = rules.get("blocking_shot_types", {})
    if isinstance(types, dict):
        return set(types.get(moment_type, types.get("broll", ["TITLE_CARD", "CUSTOM_VISUAL"])))
    return {"TITLE_CARD", "CUSTOM_VISUAL"}


def composited_layouts() -> set[str]:
    return set(load_rules().get("composited_layouts", []))


def hook_composited_ignore_blocking() -> set[str]:
    return set(load_rules().get("hook_composited_ignore_blocking", []))


def hook_frame_limit() -> int:
    return int(load_rules().get("hook_frame_limit", 90))
